### PR TITLE
TMA-299: Data product used in bricks

### PR DESCRIPTION
**Pull request by kubamahnert**

### CODE REVIEW

File: lib/gooddata/lcm/actions/apply_custom_maql.rb [hunk 1]
```diff
@@ -44,9 +44,10 @@ def call(params)
 
           domain_name = params.organization || params.domain
           domain = client.domain(domain_name) || fail("Invalid domain name specified - #{domain_name}")
+          data_product = params.data_product
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/apply_custom_maql.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/actions/associate_clients.rb [hunk 2]
```diff
@@ -45,14 +45,23 @@ def call(params)
...
-            segment = domain.segments(segment_name)
+            segment = domain.segments(segment_name, data_product)
+            segments[segment_name] = segment
             (clients.map(&:id) - segment.clients.map(&:id)).each do |c|
               segment.create_client(id: c)
             end
           end
 
+          params.clients.map! do |data|
+            data[:data_product_id] = segments[data[:segment]].data_product.data_product_id
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/associate_clients.rb]: Calls 'data[:segment]' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/associate_clients.rb [hunk 3]
```diff
@@ -45,14 +45,23 @@ def call(params)
...
           params.clients.group_by { |data| data[:segment] }.each do |segment_name, clients|
-            segment = domain.segments(segment_name)
+            segment = domain.segments(segment_name, data_product)
+            segments[segment_name] = segment
             (clients.map(&:id) - segment.clients.map(&:id)).each do |c|
               segment.create_client(id: c)
             end
           end
 
+          params.clients.map! do |data|
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/associate_clients.rb]: Calls 'params.clients' 4 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/associate_clients.rb [hunk 4]
```diff
@@ -45,14 +45,23 @@ def call(params)
 
           domain_name = params.organization || params.domain
           domain = client.domain(domain_name) || fail("Invalid domain name specified - #{domain_name}")
+          data_product = params.data_product
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/associate_clients.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/actions/collect_client_projects.rb [hunk 5]
```diff
@@ -33,7 +33,9 @@ def call(params)
           client = params.gdc_gd_client
 
           domain = client.domain(params.organization) || fail("Invalid domain name specified - #{params.organization}")
-          all_segments = domain.segments
+
+          data_product = params.data_product
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/collect_client_projects.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/actions/collect_data_product.rb [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        param :data_product, instance_of(Type::StringType), required: false
+      end
+
+      RESULT_HEADER = [
+        :data_product
+      ]
+
+      class << self
+        def call(params)
+          params = params.to_hash
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/collect_data_product.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/actions/collect_data_product.rb [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+#
+# Copyright (c) 2010-2017 GoodData Corporation. All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+require_relative 'base_action'
+
+module GoodData
+  module LCM2
+    class CollectDataProduct < BaseAction
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/collect_data_product.rb]: Has no descriptive comment (IrresponsibleModule)

File: lib/gooddata/lcm/actions/collect_data_product.rb [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        description 'DataProduct to manage'
+        param :data_product, instance_of(Type::StringType), required: false
+      end
+
+      RESULT_HEADER = [
+        :data_product
+      ]
+
+      class << self
+        def call(params)
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/collect_data_product.rb]: Has approx 12 statements (TooManyStatements)

File: lib/gooddata/lcm/actions/collect_data_product.rb [hunk 9]
```diff
@@ -0,0 +1,62 @@
+# encoding: UTF-8
+#
+# Copyright (c) 2010-2017 GoodData Corporation. All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+require_relative 'base_action'
+
+module GoodData
+  module LCM2
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/collect_data_product.rb]: Has the name 'lcm2' (UncommunicativeModuleName)

File: lib/gooddata/lcm/actions/collect_segment_clients.rb [hunk 10]
```diff
@@ -43,7 +43,8 @@ def call(params)
 
           domain_name = params.organization || params.domain
           domain = client.domain(domain_name) || fail("Invalid domain name specified - #{domain_name}")
-          domain_segments = domain.segments
+          data_product = params.data_product
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/collect_segment_clients.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/actions/collect_segments.rb [hunk 11]
```diff
@@ -24,21 +24,18 @@ class CollectSegments < BaseAction
...
         def call(params)
-          client = params.gdc_gd_client
-
-          domain_name = params.organization || params.domain
-          domain = client.domain(domain_name) || fail("Invalid domain name specified - #{domain_name}")
-          domain_segments = domain.segments
-          params.gdc_logger.info("Domain segments: #{domain_segments}")
+          data_product = params.data_product
+          data_product_segments = data_product.segments
+          params.gdc_logger.info("Domain segments: #{data_product_segments}")
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/collect_segments.rb]: Calls 'params.gdc_logger' 3 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/collect_segments.rb [hunk 12]
```diff
@@ -24,21 +24,18 @@ class CollectSegments < BaseAction
 
       class << self
         def call(params)
-          client = params.gdc_gd_client
-
-          domain_name = params.organization || params.domain
-          domain = client.domain(domain_name) || fail("Invalid domain name specified - #{domain_name}")
-          domain_segments = domain.segments
-          params.gdc_logger.info("Domain segments: #{domain_segments}")
+          data_product = params.data_product
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/collect_segments.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/actions/create_segment_masters.rb [hunk 13]
```diff
@@ -48,7 +48,8 @@ def call(params)
 
           domain_name = params.organization || params.domain
           domain = client.domain(domain_name) || fail("Invalid domain name specified - #{domain_name}")
-          domain_segments = domain.segments
+          data_product = params.data_product
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/create_segment_masters.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/actions/ensure_data_product.rb [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+      class << self
+        def call(params)
+          params = params.to_hash
+          client = params.gdc_gd_client
+          domain_name = params.organization || params.domain
+          domain = client.domain(domain_name) || fail("Invalid domain name specified - #{domain_name}")
+
+          if params.key?(:data_product)
+            begin
+              data_product = domain.data_products(params.data_product)
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/ensure_data_product.rb]: Calls 'params.data_product' 3 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/ensure_data_product.rb [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+        param :data_product, instance_of(Type::StringType), required: false
+      end
+
+      RESULT_HEADER = [
+        :data_product
+      ]
+
+      class << self
+        def call(params)
+          params = params.to_hash
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/ensure_data_product.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/actions/ensure_data_product.rb [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+#
+# Copyright (c) 2010-2017 GoodData Corporation. All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+require_relative 'base_action'
+
+module GoodData
+  module LCM2
+    class EnsureDataProduct < BaseAction
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/ensure_data_product.rb]: Has no descriptive comment (IrresponsibleModule)

File: lib/gooddata/lcm/actions/ensure_data_product.rb [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+        description 'DataProduct to ensure'
+        param :data_product, instance_of(Type::StringType), required: false
+      end
+
+      RESULT_HEADER = [
+        :data_product
+      ]
+
+      class << self
+        def call(params)
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/ensure_data_product.rb]: Has approx 8 statements (TooManyStatements)

File: lib/gooddata/lcm/actions/ensure_data_product.rb [hunk 18]
```diff
@@ -0,0 +1,53 @@
+# encoding: UTF-8
+#
+# Copyright (c) 2010-2017 GoodData Corporation. All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+require_relative 'base_action'
+
+module GoodData
+  module LCM2
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/ensure_data_product.rb]: Has the name 'lcm2' (UncommunicativeModuleName)

File: lib/gooddata/lcm/actions/provision_clients.rb [hunk 19]
```diff
@@ -33,16 +33,16 @@ class ProvisionClients < BaseAction
...
           domain_name = params.organization || params.domain
           domain = client.domain(domain_name) || fail("Invalid domain name specified - #{domain_name}")
 
-          synchronize_projects = []
-
           begin
-            results = params.segments.pmap do |segment|
-              tmp = domain.provision_client_projects(segment.segment_id).pmap do |m|
+            results = params.segments.map do |segment|
+              segment_object = domain.segments(segment.segment_id, data_product)
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/provision_clients.rb]: Calls 'segment.segment_id' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/provision_clients.rb [hunk 20]
```diff
@@ -33,16 +33,16 @@ class ProvisionClients < BaseAction
 
       class << self
         def call(params)
+          synchronize_projects = []
+          data_product = params.data_product
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/provision_clients.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/actions/provision_clients.rb [hunk 21]
```diff
@@ -33,16 +33,16 @@ class ProvisionClients < BaseAction
...
           domain = client.domain(domain_name) || fail("Invalid domain name specified - #{domain_name}")
 
-          synchronize_projects = []
-
           begin
-            results = params.segments.pmap do |segment|
-              tmp = domain.provision_client_projects(segment.segment_id).pmap do |m|
+            results = params.segments.map do |segment|
+              segment_object = domain.segments(segment.segment_id, data_product)
+              tmp = segment_object.provision_client_projects.map do |m|
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/provision_clients.rb]: Contains iterators nested 2 deep (NestedIterators)

Comment (yenkins) [inline lib/gooddata/lcm/actions/provision_clients.rb]: Has the variable name 'm' (UncommunicativeVariableName)

File: lib/gooddata/lcm/actions/segments_filter.rb [hunk 22]
```diff
@@ -29,9 +29,16 @@ def call(params)
           fail "Parameter 'segments' contains duplicate segment id(s): #{duplicated_segment_ids.join(', ')}" if duplicated_segment_ids.any?
 
           if params.segments_filter
+            data_product = params.data_product
+            data_product_segments_ids = data_product.segments.map(&:segment_id)
+
+            filtered_segments = filtered_segments.select do |segment|
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/segments_filter.rb]: Calls 'filtered_segments.select' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/segments_filter.rb [hunk 23]
```diff
@@ -29,9 +29,16 @@ def call(params)
...
+              data_product_segments_ids.include?(segment.segment_id.downcase)
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/segments_filter.rb]: Calls 'segment.segment_id' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/lcm/actions/segments_filter.rb]: Calls 'segment.segment_id.downcase' 2 times (DuplicateMethodCall)

File: lib/gooddata/lcm/actions/segments_filter.rb [hunk 24]
```diff
@@ -29,9 +29,16 @@ def call(params)
           fail "Parameter 'segments' contains duplicate segment id(s): #{duplicated_segment_ids.join(', ')}" if duplicated_segment_ids.any?
 
           if params.segments_filter
+            data_product = params.data_product
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/segments_filter.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/actions/synchronize_clients.rb [hunk 25]
```diff
@@ -45,7 +45,8 @@ def call(params)
 
           domain_name = params.organization || params.domain
           domain = client.domain(domain_name) || fail("Invalid domain name specified - #{domain_name}")
-          domain_segments = domain.segments
+          data_product = params.data_product
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_clients.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/actions/synchronize_etls_in_segment.rb [hunk 26]
```diff
@@ -40,14 +40,13 @@ class SynchronizeETLsInSegment < BaseAction
...
-          domain_name = params.organization || params.domain
-          domain = client.domain(domain_name) || fail("Invalid domain name specified - #{domain_name}")
+          data_product = params.data_product
           synchronize_segments = params.synchronize.group_by do |info|
             info[:segment_id]
           end
 
           results = synchronize_segments.pmap do |segment_id, synchronize|
-            segment = domain.segments(segment_id)
+            segment = data_product.segments.find { |s| s.segment_id == segment_id }
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_etls_in_segment.rb]: Contains iterators nested 2 deep (NestedIterators)

Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_etls_in_segment.rb]: Has the variable name 's' (UncommunicativeVariableName)

File: lib/gooddata/lcm/actions/synchronize_new_segments.rb [hunk 27]
```diff
@@ -28,7 +28,8 @@ def call(params)
 
           domain_name = params.organization || params.domain
           domain = client.domain(domain_name) || fail("Invalid domain name specified - #{domain_name}")
-          domain_segments = domain.segments
+          data_product = params.data_product
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_new_segments.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/models/client.rb [hunk 28]
```diff
@@ -39,7 +39,9 @@ def [](id, opts = {})
         fail ArgumentError, 'No client specified' if client.nil?
 
         if id == :all
-          tenants_uri = domain.segments_uri + "/clients?segment=#{CGI.escape(segment.segment_id)}"
+          data_product = segment.data_product
+          tenants_uri = GoodData::DataProduct::ONE_DATA_PRODUCT_PATH % { domain_name: domain.name, id: data_product.data_product_id }
```
Comment (yenkins) [inline lib/gooddata/models/client.rb]: Refers to 'domain' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/models/data_product.rb [hunk 29]
```diff
@@ -0,0 +1,149 @@
...
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+require_relative '../rest/resource'
+
+module GoodData
+  class DataProduct < Rest::Resource
+    include Mixin::UriGetter
+
+    attr_accessor :domain
```
Comment (yenkins) [inline lib/gooddata/models/data_product.rb]: Is a writable attribute (Attribute)

File: lib/gooddata/models/data_product.rb [hunk 30]
```diff
@@ -0,0 +1,149 @@
...
+    end
+
+    def delete(options = {})
+      segments.peach { |s| s.delete(options) }
+      client.delete(uri) if uri
+      self
+    end
+
+    def update_clients(data, options = {})
+      if options[:delete_extra] && options[:delete_extra_in_segments]
```
Comment (yenkins) [inline lib/gooddata/models/data_product.rb]: Calls 'options[:delete_extra]' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/models/data_product.rb]: Calls 'options[:delete_extra_in_segments]' 3 times (DuplicateMethodCall)

File: lib/gooddata/models/data_product.rb [hunk 31]
```diff
@@ -0,0 +1,149 @@
...
+                                  .join(',')
+        uri = data_products_uri + "/updateClients?deleteExtraInSegments=#{segments_to_delete_in}"
+        res = client.post(uri, updateClients: { items: payload })
+      else
+        res = client.post(data_products_uri + '/updateClients', updateClients: { items: payload })
+      end
+      data = GoodData::Helpers.get_path(res, ['updateClientsResponse'])
+      if data
+        result = data.flat_map { |k, v| v.map { |h| GoodData::Helpers.symbolize_keys(h.merge('type' => k)) } }
+        result.select { |r| r[:status] == 'DELETED' }.peach { |r| r[:originalProject] && client.delete(r[:originalProject]) } if options[:delete_projects]
```
Comment (yenkins) [inline lib/gooddata/models/data_product.rb]: Calls 'r[:originalproject]' 2 times (DuplicateMethodCall)

Comment (yenkins) [inline lib/gooddata/models/data_product.rb]: Has the variable name 'r' (UncommunicativeVariableName)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/516/ : SUCCESS in 7s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/254/ : FAILURE in 58s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/560/ : SUCCESS in 33s

**Comment by panjan** [comment]
ok to merge

**Pull request closed without merging**